specification/dartLangSpec.tex

```diff
@@ -3808,17 +3810,40 @@ \subsection{Superinterfaces}
 an enumerated type (\ref{enums}),
 a deferred type (\ref{staticTypes}), type \DYNAMIC{} (\ref{typeDynamic}),
 or type \code{FutureOr<$T$>} for any $T$ (\ref{typeFutureOr}).
-It is a compile-time error if two elements in the type list of the \IMPLEMENTS{} clause of a class $C$ specifies the same type $T$.
-It is a compile-time error if the superclass of a class $C$ is one of the elements of the type list of the \IMPLEMENTS{} clause of $C$.
-It is a compile-time error if a class $C$ has two superinterfaces that are different instantiations of the same generic class.
-\commentary{For example, a class may not have both `List<int>` and `List<num>` as superinterfaces.}
-% If we need to allow multiple instantiations, they'll need to have a most
-% specific one, and then we can add the following clause
-%, unless it implements one that is a subtype of all the other. \commentary{This ensures that each class implements one {\em most specific} version of a generic class' interface.}
+It is a compile-time error if two elements in the type list of
+the \IMPLEMENTS{} clause of a class $C$ specifies the same type $T$.
+It is a compile-time error if the superclass of a class $C$ is
+one of the elements of the type list of the \IMPLEMENTS{} clause of $C$.
 
-\rationale{
-One might argue that it is harmless to repeat a type in the superinterface list, so why make it an error? The issue is not so much that the situation described in program source is erroneous, but that it is pointless.
-As such, it is an indication that the programmer may very well have meant to say something else - and that is a mistake that should be called to her or his attention.
+\rationale{%
+One might argue that it is harmless to repeat a type in the superinterface list,
+so why make it an error?
+The issue is not so much that the situation is erroneous,
+but that it is pointless.
+As such, it is an indication that the programmer may very well have meant
+to say something else---and that is a mistake that should be
+called to her or his attention.%
+}
+
+\LMHash{}%
+It is a compile-time error if a class $C$ has two superinterfaces
+that are different instantiations of the same generic class.
+\commentary{%
+For example, a class can not have
+both \code{List<int>} and \code{List<num>} as superinterfaces,
+directly or indirectly.%
+}
+
+\LMHash{}%
+When a generic class $C$ declares a type parameter $X$,
+it is a compile-time error if $X$ occurs in a non-covariant position
+% Could say 'a direct superinterface', but it is easy to see that it is
+% enough to check direct superinterfaces, and it is then true
+% for indirect ones as well.
+in a superinterface of $C$.
```

Comment:
I think "in a superinterface" is ambiguous. It can mean in the superinterface type or in the superinterface member signatures or ....
Maybe:

> in a type parameter of a superinterface of $C$.

That's the only place it can occur, and it makes it clear(er) that we don't look through the superinterface signatures.

Reply:
Changed it to say 'in a type which specifies a superinterface'. It's dangerous to say 'in a type parameter' because this will swap the positions if the corresponding type parameter is contravariant.